trin-core/src/portalnet/overlay_service.rs

```diff
@@ -250,6 +257,12 @@ pub struct OverlayService<TContentKey, TMetric> {
     request_tx: UnboundedSender<OverlayRequest>,
     /// A map of active outgoing requests.
     active_outgoing_requests: Arc<RwLock<HashMap<OverlayRequestId, ActiveOutgoingRequest>>>,
+    /// Keeps track of which queries a given node has been requested on behalf of.
+    active_query_requests: Arc<RwLock<HashMap<NodeId, QueryId>>>,
+    /// All of the iterative queries currently being performed.
+    queries: QueryPool<QueryInfo, NodeId, Enr>,
+    /// A map of Node IDs to the most recently seen ENR for that Node ID. Todo: make this persistent.
+    node_enrs: RwLock<HashMap<NodeId, Enr>>,
```

Comment:
In order to get the ENR for a given Node ID from the routing table, I was originally doing the following :
```
let optional_node = match self.kbuckets.write().entry(&key) {
            kbucket::Entry::Present(ref mut entry, _) => Some(entry.value().clone()),
            kbucket::Entry::Pending(ref mut entry, _) => Some(entry.value().clone()),
            _ => None,
};
```

But I found that sometimes ENRs passed into `process_discovered_enrs` were not yet present in the routing table (the above approach would return `None`) by the time the `query_pool` needed a request to be sent to the node a few moments later. 

I determined (perhaps incorrectly) that the `kbucket` has a delay period before the entry could be accessed, and figured that because we intend to have a flat, persistent store for mapping node IDs to ENRs outside of the kbuckets (incase they are full) anyway, this would be a good time to introduce/use that mechanism (in a temporarily not-persistent fashion).

Let me know if you know of a reason for not finding the nodes in the routing table close-to-immediately after calling `process_discovered_enrs`  that I might have overlooked.

Reply:
I could be missing something here, but I think we resolve that issue if we update the query state at the end of `process_discovered_enrs` after all ENR updates/insertions.

If we wait until all ENR changes are made to the routing table before we update the query state, then by the time we call `on_success` for that query with the new ENRs, those ENRs are guaranteed to be in the routing table (assuming there is enough space).

Like you said, it might happen that not all of the ENRs from the query fit in the routing table. I'm not so sure that this is a problem though, at least for now. If the corresponding bucket is full, then that would mean that we already know `k` nodes who are just as close to the target as the left out ENR. Presumably, those known ENRs would be among the known closest nodes when we initiate the query. Any ENRs that don't fit in the routing table could be filtered from the query update.

Reply:
So looking into this a bit more, we could adopt [this method](https://github.com/sigp/discv5/blob/506c9afe6a037dc3ba920f691068dac29f6cba7c/src/service.rs#L499) of `sigp/discv5`. This checks both the routing table (for present entries) and all outstanding queries for a given `NodeId`. It checks the `untrusted_enrs` field of the `QueryInfo`, which we update when we perform the query update.

This seems far better than what I suggested above, which would limit the reach of the lookup procedure. With this solution, I don't see the need for `node_enrs`.